bpf/lib/nat.h

```diff
@@ -871,6 +958,66 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
 			tuple.dport = icmphdr.un.echo.id;
 			tuple.sport = 0;
 			break;
+		case ICMP_DEST_UNREACH:
+			switch (icmphdr.code) {
+			case ICMP_FRAG_NEEDED: {
+				struct iphdr iphdr;
+				__be16 identifier;
+				__u8 type;
+				__u32 icmpoff = off + sizeof(icmphdr);
+
+				/* According to the RFC 5508, any networking
+				 * equipment that is responding with an ICMP Error
+				 * packet should embed the original packet in its
+				 * response.
+				 */
+
+				if (ctx_load_bytes(ctx, icmpoff, &iphdr,
+						   sizeof(iphdr)) < 0)
+					return DROP_INVALID;
+
+				/* From the embedded IP headers we should be able
+				 * to determine corresponding protocol, IP src/dst
+				 * of the packet sent to resolve the NAT session.
+				 */
+				tuple.nexthdr = iphdr.protocol;
+				tuple.saddr = iphdr.daddr;
+				tuple.daddr = iphdr.saddr;
+
+				icmpoff += ipv4_hdrlen(&iphdr);
+				switch (tuple.nexthdr) {
+				case IPPROTO_TCP:
+				case IPPROTO_UDP:
+					if (ctx_load_bytes(ctx, icmpoff, &l4hdr,
+							   sizeof(l4hdr)) < 0)
+						return DROP_INVALID;
+					tuple.sport = l4hdr.dport;
+					tuple.dport = l4hdr.sport;
```

Comment:
It's good point, but wondering whether we could be in that situation since that would probably mean that the DF bit isn't set, right? no reasons that the networking equipment to respond with a need to fragment.

Reply:
Yep, sounds reasonable. Would be worth a comment imho, at some point we will hopefully audit all the remaining paths for missing Fragmentation support (https://github.com/cilium/cilium/issues/11180). So if we think it's not needed here, let's spell that out.

Reply:
Done, please let me know if the phrasing sounds right for you :-)